src/github.com/matrix-org/dendrite/cmd/syncserver-integration-tests/main.go

```diff
@@ -60,6 +61,23 @@ var exe = test.KafkaExecutor{
 	OutputWriter: os.Stderr,
 }
 
+var (
+	lastRequestMutex sync.Mutex
+	lastRequestErr   error
+)
+
+func setLastRequestError(err error) {
+	lastRequestMutex.Lock()
```

Comment:
`lastRequestErr` is set in the goroutine which is hammering `/sync`, but it gets read in the main goroutine when the test times out. Go has a lovely [memory model document](https://golang.org/ref/mem) where they outline all the specific subtle cases where it is safe to read/write like this and the cases where it is not. Their general advice is:

> Programs that modify data being simultaneously accessed by multiple goroutines must serialize such access.
>
> To serialize access, protect the data with channel operations or other synchronization primitives such as those in the sync and sync/atomic packages.
>
> If you must read the rest of this document to understand the behavior of your program, you are being too clever.
>
> Don't be clever.

I wholeheartedly concur and therefore treat any situation where I need to share variables across goroutines by wrapping them up in sync primitives.